### ISSUE TITLE
Validate a11y attributes

### ISSUE THREAD

**Issue by dgp1130** [angular/angular#44869]
### Which @angular/* package(s) are relevant/releated to the feature request?

compiler-cli

### Description

We should have the Angular compiler validate the types for a11y features of the DOM, such as `role` or `aria-*` attributes to make sure they are always set to valid values. This could allow the compiler to error or warn on bad a11y values and help users identify mistakes in their code which can lead to a more accessible experience.

### Proposed solution

This could potentially be done as an extended diagnostic, or possibly through a `*.d.ts` typecheck or other direct compiler integration. I'm not sure which would be more appropriate here.

### Alternatives considered

N/A

**Comment by flash-me** [comment]
Is it checking on a syntactical level or are you going to check also semantics?

e.g.  the attribute [`aria-checked`](https://www.w3.org/TR/wai-aria-1.1/#aria-checked) is not valid on `div`, `p`, `span` elements etc.

However, if a custom checkbox (or option etc.)  is implemented and is semantically defined as such by providing the [role](https://www.w3.org/TR/wai-aria-1.1/#checkbox), 
the `aria-checked` attribute becomes totally valid:

```html
<!-- custom checkbox implementation -->
<div role="checkbox" aria-checked="true"></div>
```

I mean...the possible [amount of combinations](https://www.w3.org/TR/wai-aria-1.1/img/rdf_model.svg) (and even more coming with [WAI-ARIA 1.2](https://www.w3.org/TR/wai-aria-1.2/#roles_categorization)) is huge.

Nonetheless, looking forward for this feature, even if it's only checking on the syntactical level!

cheers
flash ⚡

**Comment by dgp1130** [comment]
I think checking semantics should be on the table, an extended diagnostic should be able to identify that you out `aria-checked` on the wrong tag.

As you said, there's a lot of complexity there, so we may start small and expand the check over time. We'll have to see what is practical to implement.

**Comment by twerske** [comment]
Love this idea! A great small start would be adding a check for `alt` on `<img>` elements.

**Comment by twerske** [comment]
Additional checks to consider
- check for a `title` attribute in the `Route` definitions
- Heading levels should have a meaningful hierarchy, e.g.: (check nesting order of h1, h2 etc)
- Form labels and labelFor could likely be statically checked or at least add warnings for

**Comment by flash-me** [comment]
Would love to participate / contribute on this one

cheers
flash ⚡

**Comment by markgoho** [comment]
imo accessibility is too big a domain for Angular to provide any level of real guidance and this should best be left to other tools that already provide the same level of feedback

the worst thing about this idea is the false sense of compliance that it gives devs thinking no red squiggles means compliance

**Comment by moniuch** [comment]
> imo accessibility is too big a domain for Angular to provide any level of real guidance and this should best be left to other tools that already provide the same level of feedback
> 
> the worst thing about this idea is the false sense it gives devs thinking no red squiggles means compliance

Totally agree on that. On top of what @markgoho said:

Given that WCAG are only guidelines and not a standard/specs, it means that every team should be able to decide on how and if they want to implement them. For this there are appropriate opt-in tools, such as [angular/eslint](https://github.com/angular-eslint/angular-eslint#functionality) where individual rules can be enabled.

Also, as the guidelines will most likely evolve, this will create additional burden on the Angular project to keep pace. In case this fails, this will cause frustration, users will eventually resort to other tools focused specifically on a11y, and the contributors' effort on this issue might eventually be wasted.
